src/app/public/modules/angular-tree/angular-tree-root.component.spec.ts

```diff
@@ -0,0 +1,571 @@
+
+  // #region helpers
+  function getToolbar(): HTMLElement {
+    return document.querySelector('.sky-angular-tree-toolbar') as HTMLElement;
+  }
+
+  function getSelectAllButton(): HTMLElement {
+    return document.querySelector('.sky-angular-tree-select-all-btn') as HTMLElement;
+  }
+
+  function getClearAllButton(): HTMLElement {
+    return document.querySelector('.sky-angular-tree-clear-all-btn') as HTMLElement;
+  }
+
+  function getSkyCheckboxes(): NodeListOf<HTMLElement> {
+    return document.querySelectorAll('sky-checkbox');
+  }
+
+  function getCheckboxInputs(): NodeListOf<HTMLInputElement> {
+    return document.querySelectorAll('sky-checkbox input');
+  }
+
+  function getExpandButton(): HTMLElement {
+    return document.querySelector('.sky-angular-tree-expand-all-btn') as HTMLElement;
+  }
+
+  function getCollapseButton(): HTMLElement {
+    return document.querySelector('.sky-angular-tree-collapse-all-btn') as HTMLElement;
+  }
+
+  function getNodeContents(): NodeListOf<HTMLElement> {
+    return document.querySelectorAll('.node-content-wrapper');
+  }
+
+  function getNodeWrappers(): NodeListOf<HTMLElement> {
+    return document.querySelectorAll('.node-wrapper');
+  }
+
+  function clickSelectAll(): void {
+    getSelectAllButton().click();
+    tick();
+    fixture.detectChanges();
+  }
+
+  function clickClearAll(): void {
+    getClearAllButton().click();
+    tick();
+    fixture.detectChanges();
+  }
+
+  function clickExpand(): void {
+    getExpandButton().click();
+    tick();
+    fixture.detectChanges();
+  }
+
+  function clickCollapse(): void {
+    getCollapseButton().click();
+    tick();
+    fixture.detectChanges();
+  }
+
+  // Selection helpers
+  // Note: checking for checkboxes and nodes should be separate. You may not have a checkbox for every node.
+  // For example: Leaf-node only mode will hide checkboxes for parents.
+  // expectNodeToBeSelected() will check if the node has proper styles and the tree model is updated correctly.
+  // expectCheckboxToBeChecked() will check if our custom implementation of sky-checkbox has the proper checked state.
+
+  // Use 1-based indexes!
+  function expectNodeToBeSelected(nodeIndex: number, selected: boolean): void {
+    const nodeWrappers = getNodeWrappers();
+
+    if (selected) {
+      expect(nodeWrappers[nodeIndex - 1]).toHaveCssClass('sky-angular-tree-node-selected');
+      expect(component.selectedLeafNodeIds[nodeIndex]).toEqual(true);
+    } else {
+      expect(nodeWrappers[nodeIndex - 1]).not.toHaveCssClass('sky-angular-tree-node-selected');
+      expect(component.selectedLeafNodeIds[nodeIndex] || false).toEqual(false);
+    }
+  }
+
+  // Use 1-based indexes!
+  function expectCheckboxToBeChecked(nodeIndex: number, selected: boolean): void {
+    const checkboxInputs = getCheckboxInputs();
+    expect(checkboxInputs[nodeIndex - 1].checked).toEqual(selected);
+  }
+
+  function setupCascadingMode(): void {
```

Comment:
Yes, because some tests require a custom configuration to the tree. The `detectChanges()` will draw the tree, so I wanted to prevent that from automatically happening in each test. I wanted the setup to be deliberate in each one.

Reply:
Makes sense